swift/apple/FirezoneKit/Sources/FirezoneKit/Helpers/LogCompressor.swift

```diff
@@ -9,91 +9,116 @@ import AppleArchive
+///
+/// Once the writeStream is opened, the remaining operations are the same for both.
+public struct LogCompressor {
+  enum CompressionError: Error {
+    case unableToReadSourceDirectory
+    case unableToOpenWriteStream
+    case unableToOpenCompressionStream
+    case unableToOpenEncodeStream
+    case unableToDefineHeaderKeys
   }
 
-  public func compressFolder(destinationURL: URL? = nil) async throws {
-    try await compressFolderReturningURL(destinationURL: destinationURL)
-  }
+  public init() {}
 
-  @discardableResult
-  public func compressFolderReturningURL(destinationURL: URL? = nil) async throws -> URL? {
-    guard let logFilesFolderURL = SharedAccess.logFolderURL,
-      let logFilesFolderPath = FilePath(logFilesFolderURL)
+  public func start(
+    source directory: URL,
+    to file: URL
+  ) throws {
+    guard let path = FilePath(file),
+          let stream = ArchiveByteStream.fileStream(
+            path: path,
+            mode: .writeOnly,
+            options: [.create],
+            permissions: FilePermissions(rawValue: 0o644)
+          )
     else {
-      throw SettingsViewError.logFolderIsUnavailable
+      throw CompressionError.unableToOpenWriteStream
     }
 
-    let fileManager = FileManager.default
-    let fileURL =
-      destinationURL
-      ?? fileManager.temporaryDirectory.appendingPathComponent(fileName)
-
-    // Remove logfile if it happens to exist
-    try? fileManager.removeItem(at: fileURL)
-
-    // Create the file stream to write the compressed file
-    guard let filePath = FilePath(fileURL),
-      let writeFileStream = ArchiveByteStream.fileStream(
-        path: filePath,
-        mode: .writeOnly,
-        options: [.create],
-        permissions: FilePermissions(rawValue: 0o644))
+    try compress(source: directory, writeStream: stream)
+  }
+
+  public func start(
+    source directory: URL,
+    to byteStream: TunnelArchiveByteStream
+  ) throws {
+    guard let stream = ArchiveByteStream.customStream(
+      instance: byteStream
+    )
     else {
-      Log.app.error("\(#function): Couldn't create the file stream")
-      return nil
-    }
-    defer {
-      try? writeFileStream.close()
+      throw CompressionError.unableToOpenWriteStream
     }
 
-    // Create the compression stream
-    guard
-      let compressStream = ArchiveByteStream.compressionStream(
-        using: .lzfse,
-        writingTo: writeFileStream)
+    try compress(source: directory, writeStream: stream)
+  }
+
+  // Compress to a given writeStream which was opened either from a FilePath or
+  // TunnelArchiveByteStream
+  private func compress(
+    source directory: URL,
+    writeStream: ArchiveByteStream
+  ) throws {
+    let compressionStream = try openCompressionStream(writeStream)
+    let encodeStream = try openEncodeStream(compressionStream)
+    let keySet = try defineHeaderKeys()
+
+    guard let sourcePath = FilePath(directory)
```

Comment:
Similarly here, should we defer it to the caller to pass the correct datatype (a `FilePath`) to begin with?